packages/backend/src/services/SpaceService/SpaceService.ts

```diff
@@ -257,62 +260,126 @@ export class SpaceService extends BaseService {
         return updatedSpace;
     }
 
-    async moveSpace(
-        user: SessionUser,
-        spaceUuid: string,
-        parentSpaceUuid: string | null,
+    private async hasAccess(
+        action: AbilityAction,
+        actor: {
+            user: SessionUser;
+            projectUuid: string;
+        },
+        resource: {
+            spaceUuid: string;
+            newParentSpaceUuid?: string;
+        },
     ) {
-        const checkSpace = async (uuid: string) => {
-            const space = await this.spaceModel.getSpaceSummary(uuid);
-            const spaceAccess = await this.spaceModel.getUserSpaceAccess(
-                user.userUuid,
-                uuid,
+        const space = await this.spaceModel.getSpaceSummary(resource.spaceUuid);
+        const spaceAccess = await this.spaceModel.getUserSpaceAccess(
+            actor.user.userUuid,
+            space.parentSpaceUuid ?? resource.spaceUuid, // if newParentSpaceUuid is not provided, use the spaceUuid
+        );
+
+        const isActorAllowedToPerformAction = actor.user.ability.can(
+            action,
+            subject('Space', {
+                organizationUuid: actor.user.organizationUuid,
+                projectUuid: actor.projectUuid,
+                isPrivate: space.isPrivate,
+                access: spaceAccess,
+            }),
+        );
+
+        if (!isActorAllowedToPerformAction) {
+            throw new ForbiddenError(
+                `You don't have access to ${action} this space`,
             );
-            if (
-                user.ability.cannot(
-                    'manage',
+        }
+
+        if (resource.newParentSpaceUuid) {
```

Comment:
if `newParentSpaceUuid === null` should you still check if user can create space in general?